### PR TITLE
HOTFIX: avoid placement of unnecessary transient standby tasks

### PR DESCRIPTION
**Pull request by ableegoldman**
Minor fix to avoid creating unnecessary standby tasks, especially when these may be surprising or unexpected as in the case of an application with `num.standby.replicas = 0` and warmup replicas disabled.

The "bug" here was introduced during the fix for an issue with cooperative rebalancing and in-memory stores. The fundamental problem is that in-memory stores cannot be unassigned from a consumer for any period, however temporary, without being closed and losing all the accumulated state. This caused some grief when the new HA task assignor would assign an active task to a node based on the readiness of the standby version of that task, but would have to remove the active task from the initial assignment so it could first be revoked from its previous owner, as per the cooperative rebalancing protocol. This temporary gap in any version of that task among the consumer's assignment for that one intermediate rebalance would end up causing the consumer to lose all state for it, in the case of in-memory stores.

To fix this, we simply began to place standby tasks on the intended recipient of an active task awaiting revocation by another consumer. However, the fix was a bit of an overreach, as we assigned these temporary standby tasks in all cases, regardless of whether there had previously been a standby version of that task. We can narrow this down without sacrificing any of the intended functionality by only assigning this kind of standby task where the consumer had previously owned some version of it that would otherwise potentially be lost.

### CODE REVIEW

**Comment by ableegoldman** [comment]
call for review anyone cc @guozhangwang @cadonna @mjsax @lihaosky

**Comment by ableegoldman** [comment]
Test failures are unrelated

Super small fix, would like to get it into 3.6 though

ping also @vvcephei @showuon @wcarlson5 for a review

**Review by bbejeck — COMMENTED** [review]
@ableegoldman I'm not super familiar with the code but your explanation and the change seem straightforward enough, so it overall looks good - can you add a test case for this fix?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java [hunk 1]
```diff
@@ -1063,9 +1071,13 @@ private Map<TaskId, Set<TopicPartition>> buildStandbyTaskMap(final String consum
         }
 
         for (final TaskId task : revokedTasks) {
-            if (allStatefulTasks.contains(task)) {
```
Comment (lihaosky) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java:1084]: In case of rocksdb state store, will this add standby and restore state so that when active is assigned, there's more state? So it doesn't need to be a previous standby? Or this is not the intention of this code?

Another noob question, how does standby work with in-memory state store? Is it also reading from changelog topic? If so, what's the problem of losing state for in-memory state store as it can be restored from changelog topic?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java [hunk 2]
```diff
@@ -618,18 +618,24 @@ private boolean assignTasksToClients(final Cluster fullMetadata,
...
 
         final Set<TaskId> allTasks = partitionsForTask.keySet();
         statefulTasks.addAll(changelogTopics.statefulTaskIds());
 
-        log.debug("Assigning tasks {} including stateful {} to clients {} with number of replicas {}",
-            allTasks, statefulTasks, clientStates, numStandbyReplicas());
+        log.info("Assigning stateful tasks: {}\n"
+                     + "and stateless tasks: {}",
+                 statefulTasks,
+                 allTasks.removeAll(statefulTasks));
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java]: Won't this change the content of `allTasks`? Also `removeAll` returns a boolean.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java (hunk 1 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java:1084]: Not sure I fully understand the change here either. Would appreciate some more explanations --- specifically, could you give a simple example in the PR description?

Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java:1084]: > Another noob question, how does standby work with in-memory state store? Is it also reading from changelog topic?

Yes.

>  If so, what's the problem of losing state for in-memory state store as it can be restored from changelog topic?

The problem is, that we would have to restore from the changelog, and we want to avoid this restore. In the end, there is a task that as the full state in-memory, and we would close the task, drop the state, and re-create the same task and restore. We basically want to avoid the "close and re-create" part. (For RocksDB it's not an issue to close a task, because when we re-open it, the state is still on disk and restoring from the changelog won't be necessary).

**Comment by mjsax** [comment]
> This caused some grief when the new HA task assignor would assign an active task to a node based on the readiness of the standby version of that task, but would have to remove the active task from the initial assignment so it could first be revoked from its previous owner, as per the cooperative rebalancing protocol.

Not sure if I understand this part? If we want to move a task, why would it be a problem to remove it (ie, the active one) from the assignment? Or are you saying that "removing it from the assignment" mean, _all_ "copies" of a task, ie, all active/standbies/warmups where removed?

> To fix this, we simply began to place standby tasks on the intended recipient of an active task awaiting revocation by another consumer.

Are you saying we try to promote a "warmup" as "standby" to keep the task alive?

> However, the fix was a bit of an overreach, as we assigned these temporary standby tasks in all cases, regardless of whether there had previously been a standby version of that task

Are you saying we are assigning a standby to all client, even if they did not have a corresponding warm-up tasks that we want to keep alive?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java (hunk 1 above)
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java:1084]: > Not sure I fully understand the change here either. Would appreciate some more explanations --- specifically, could you give a simple example in the PR description?

Also not 100% sure -- I left another comment -- however I believe, we current assign standbies out of think air, even into client that should not have them, but we only want these standbies for certain client that had such a task in the previous assignment. But details are also a little bit unclear to me.

**Comment by ableegoldman** [comment]
Here's an example (or two). Hopefully this clears things up. Assume two nodes and focus on task `0_1` which is stateful but in-memory. Let's start with this first case where everything works as intended:

0. Previous assignment: node A has the active task and node B has the standby for 0_1

1. TaskAssignor moves the active 0_1 to node B, assigns standby 0_1 to node A
2. When processing the `ClientState` assignments, the StreamsPartitionAssignor removes the active 0_1 from node B because it must be revoked by node A (ie cooperative protocol) -- the active 0_1 will be assigned to node B in an immediate followup rebalance
3. The StreamsPartitionerAssignor sees that it has removed an active copy of a stateful task `0_1` from node B, and then assigns a standby copy of `0_1` in its place

This example shows the mechanism working as intended: given 0 & 2, then 3 ie we want to assign a "stand-in" standby task so that there is no assignment, however temporary, where node B has _no_ copy of task 0_1 -- because otherwise, it'll close the task (rather than recycle it) and lose the in-memory state completely in between revoking the standby task and assigning the active task in the followup rebalance

The "bug" (or inconvenience, depending on your perspective) is that we actually assign this temporary "stand-in" standby task any time we see that an active version of the task was (temporarily) removed from the assignment according to the cooperative protocol. However, this is only necessary to do when that node previously had a standby copy of that state, which would be lost if no copy of 0_1 were assigned to the node in the current rebalance.

In other words, the example that doesn't quite work as intended is where all the same steps 1-3 are followed in the above case, but node B had no copy of 0_1 (standby or otherwise) in the previous assignment.

Thus, the fix is just to check whether the previous assignment had a standby copy of that task before we assign this temporary "stand-in" standby task.

It's not a correctness issue, but there are some systems which assume there are absolutely no standby tasks and would be surprised/bugged to get this surprise standby

Does that answer the above questions?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java (hunk 2 above)
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java]: ah yeah, my bad, changed to use `stream().filter()` instead

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java (hunk 1 above)
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java:1084]: Left a comment with an example of how this is supposed to work, and where it doesn't quite work as intended (ie the case being addressed by this fix). Hope this helps

**Comment by ableegoldman** [comment]
Test failures are unrelated

@bbejeck @guozhangwang @mjsax @showuon @cadonna @vvcephei  

Can someone give this a final pass when you have a few minutes? Would really like to get this (and especially the logging improvements) in before the 3.6 code freeze tomorrow.

(Btw, see my previous comment for a more in-depth explanation with examples)

**Review by wcarlson5 — APPROVED** [review]
This looks good to me! Thanks for the PR Sophie

**Comment by ableegoldman** [comment]
Thanks! Merged to trunk and cherrypicked to 3.6 & 3.5
